README.md
# Content
* [VirtualBox VM automation in Python](#about)
* [Downloads](#downloads)
* [Usage](#usage)
* [Host configuration](#host-configuration)
* [Guest configuration](#guest-configuration)
* [TODO (1.0)](#todo-10)
* [TODO (2.0)](#todo-20)
* [Example videos](#example-videos)
* [Useful links](#useful-links)
* [Donations](#donations)

# About
Python script that can be used to automate dynamic testing of binaries/scripts/documents on VMs (currently only VirtualBox is supported).
Based on VBoxManage command-line interface and does not require VirtualBox SDK.

Both Windows and Linux are tested as host OS. May work on other platforms, supported by VirtualBox and Python.

# Downloads
Latest stable version and Windows binaries are available in <a href="https://github.com/Pernat1y/vm-automation/releases" target="_blank">Releases</a>.

# Usage:
Essential commands:
```
python demo_cli.py \
    file.exe \
    --vms windows10 windows8 windows7 \
    --snapshots firefox chrome ie
```

All options (AKA --help):
```
Optional arguments:
  -h, --help            show this help message and exit

Required options:
  file                  Path to file
  --vms [VMS ...], -v [VMS ...]
                        Space-separated list of VMs to use
  --snapshots [SNAPSHOTS ...], -s [SNAPSHOTS ...]
                        Space-separated list of snapshots to use

Main options:
  --vboxmanage [VBOXMANAGE]
                        Path to vboxmanage binary (default: vboxmanage)
  --check_version       Check for latest VirtualBox version online (default: False)
  --timeout [TIMEOUT]   Timeout in seconds for both commands and VM (default: 60)
  --delay [DELAY]       Delay in seconds before/after starting VMs (default: 7)
  --threads [{0,1,2,3,4,5,6,7,8}]
                        Number of concurrent threads to run (0=number of VMs, default: 2)
  --verbosity [{debug,info,error,off}]
                        Log verbosity level (default: info)
  --debug               Print all messages. Alias for "--verbosity debug" (default: False)
  --log [LOG]           Path to log file (default: None) (console)
  --report              Generate html report (default: False)
  --record              Record video of guest' screen (default: False)
  --pcap                Enable recording of VM's traffic (default: False)
  --memdump             Dump memory VM (default: False)
  --no_time_sync        Disable host-guest time sync for VM (default: False)

VM options:
  --ui [{1,0,gui,headless}]
                        Start VMs in GUI or headless mode (default: gui)
  --login [LOGIN], --user [LOGIN]
                        Login for guest OS (default: user)
  --password [PASSWORD]
                        Password for guest OS (default: 12345678)
  --remote_folder [{desktop,downloads,documents,temp}]
                        Destination folder in guest OS to place file. (default: desktop)
  --open_with [OPEN_WITH]
                        Absolute path to app, which will open main file (default: %windir%\Explorer.exe)
  --network [{on,off}]  State of network adapter of guest OS (default: None)
  --resolution [RESOLUTION]
                        Screen resolution for guest OS. Can be set to "random" (default: None)
  --mac [MAC]           Set MAC address for guest OS. Can be set to "random" (default: None)
  --get_file [GET_FILE]
                        Get specific file from guest OS before stopping VM (default: None)
  --pre [PRE]           Script to run before main file (default: None)
  --post [POST]         Script to run after main file (default: None)
```

# Host configuration
* Both Windows and Linux are tested as host OS. May work on other platforms, supported by VirtualBox.
* You need VirtualBox (the newer the better). Proprietary Oracle VM VirtualBox Extension Pack is *not* required.
* It is recommended to have folder with 'vboxmanage' binary in the environment variables.
* You need a recent version of Python (3.7+).

# Guest configuration
* You must have Windows as the guest OS with auto login configured (or have a snapshot with a user logged in).
* You must have VirtualBox guest additions installed.
* It is strongly recommended to have live snapshots to restore to (otherwise it will be *much* slower).
* VM disk encryption is *not* supported (VBoxManage limitation).

# TODO (1.0):
* Small improvements.
* Code optimization and fixes.

# TODO (2.0):
* Use VirtualBox API.
* Distribute workload to multiple physical hosts.
* Implement web interface.
* Add option to use pre-running VMs.
* Add global progress status.
* Vagrant integration (maybe).
* VMware support (maybe).
* Code optimization and fixes.
* Better tests coverage.

# Example videos
* Windows host (version 0.9.2):

<a href="http://www.youtube.com/watch?feature=player_embedded&v=t6AWew06rxo" target="_blank"><img src="http://img.youtube.com/vi/t6AWew06rxo/0.jpg" width="320" height="240" border="10" /></a>

* Linux host (version 0.7.3):

<a href="http://www.youtube.com/watch?feature=player_embedded&v=QnXfmPVbmlo" target="_blank"><img src="http://img.youtube.com/vi/QnXfmPVbmlo/0.jpg" width="320" height="240" border="10" /></a>

# Useful links
<a href="https://github.com/hfiref0x/VBoxHardenedLoader" target="_blank">VirtualBox Hardened VM detection mitigation loader - VBoxHardenedLoader</a>

# Changelog
Version 0.11:
* Added '--file_args' option to pass an argument to the main file/executable.
* '--uac_parent' option renamed to '--open_with' as it may be used with any type of files, not only the executables.
* Updated vm_record() - added new options and updated defaults.
* Small improvements in html report.

For complete changelog see <a href="CHANGELOG.md" target="_blank">CHANGELOG.md</a>

# Donations
You can support further development with a donation (Thanks!).
* BTC: bc1q5wzj6qa3d7vtw9cehftt7gvswr60kgfgeu98z6
* BTC (Legacy): 1GDy6seYwiK92XAyoQsSeMf2LMR9pCpkY8



CHANGELOG.md
# Changelog
Version 0.11:
* Added '--file_args' option to pass an argument to the main file/executable.
* '--uac_parent' option renamed to '--open_with' as it may be used with any type of files, not only the executables.
* Updated vm_record() - added new options and updated defaults.
* Small improvements in html report.

Version 0.10.3:
* Added function vm_functions.vm_disable_time_sync() and option to disable host-guest time sync ('--no_time_sync').

Version 0.10.2:
* Added option to check VirtualBox version online (--check_version). Will warn if outdated version is used.
Windows binary does not have this option.
* Fixed '--log' option.

Version 0.10.1:
* Added function vm_functions.vm_memdump() and option to dump VM memory (--memdump).
This may take some time and disk space.
* Added script to build Windows binary using [Nuitka](https://nuitka.net/). See 'scripts' folder.

Version 0.10:
* Added option to dump all VM's network traffic to the file ('--pcap'). File will be saved as {vm_name}_{snapshot}.pcap.
* Added option '--get_file' to download file (memory dumps, logs, reports, etc) before stopping VM.
* Removed option 'keep' for all of the arguments. Omitting argument will do the same.
* Added function vm_set_mac(vm, mac) and option to change MAC address for VM before start.
Can be set to specific MAC ('--mac 80AABBCCDDEE'), 'new' for random mac in VirtualBox range or 'random' for random one.
* Logs levels tweaked to make default output more clean.
* Added argument '--debug' (alias for '--verbosity debug').
* Added standalone Windows binary. See releases section for download.

Version 0.9.2:
* Added option '--delay' to control delay before/after starting VMs. Depends on hardware performance.
* Added md5 checksum calculation.
* Updated html report generation logic.
* Additional check to make sure that number of thread <= number of VMs.

Version 0.9.1:
* Added check for original file at the end of analysis.
* Added more data to html report.
* Small cleanups.

Version 0.9:
* Added option to generate html report ('--report'). Result (including screenshots) will be saved under ./reports/<file_hash> directory.
* Added option to record video from guest VM ('--record output.webm').
* uac_parent is now applied to both pre_exec and post_exec commands too.
* Removed '--info' and '--uac_fix' arguments. Both options are enabled from now on.
* Added alias '--user' for '--login' option.

Version 0.8.2:
* Added vm_backup(vm) function. Takes live snapshot with name in 'backup_YYYY_MM_DD_HH_MM_SS' format.

Version 0.8.1:
* Fixed bug with some command line args were ignored (--ui).

Version 0.8:
* Added options 'uac_fix' and 'uac_parent' for vm_exec() function.
Used to circumvent VirtualBox error VERR_PROC_ELEVATION_REQUIRED, when trying to execute application with mandatory UAC elevation.
* Added short aliases for '--vms' and '--snapshots' options: '-v' and '-s'.
* Added option 'dictionary' for list_vms() function. List of VMs will be returned as {'vm': '/group'} dictionary.
* Added function to clone VM - vm_clone().
* Added function to record guest screen - vm_record().
* Added option to restore current snapshot: vm_snapshot_restore('vm', 'restorecurrent').
* Updated vm_screenshot() function - code to generate image index moved to main script.
* All files converted to Unix EOL (LF).

Version 0.7.3:
* Docstring added for all the functions (Sphinx format). Usage: help(function_name).
* Added functions to export and import VMs - vm_export() and vm_import(). Have separate timeout settings as operations are time consuming.

Version 0.7.2:
* Added function to enumerate guest properties - vm_enumerate().
* Updated list_ips() function.
* Added functions to control snapshots - vm_snapshot_take() and vm_snapshot_remove(). Currently not used.

Version 0.7.1:
* Added option to redirect log to file ('--log vm_automation.log'). Default: log to console ('--log console').

Version 0.7:
* Added option to control number of concurrently running tasks ('--threads 2'). Set to '0' to set to number of VMs.
* Added option to control log verbosity ('--verbosity debug|info|error').
* Added parameter ignore_status_error to vm_stop() function. May be used when trying to stop already stopped VM.
* Added aliases for vm_copyto() and vm_copyfrom() functions - vm_upload() and vm_download().
* Fixed command line arguments processing.
* Added unittests for some of the functions ('tests/test.py').
* Other minor updates for few functions.

Version 0.6.2:
* Bug fixes in list_snapshots() function.

Version 0.6.1:
* All vm_functions are now returning exactly 3 values: [exit_code, stdout, stderr].
* New function to get IP addresses of guest - vm_functions.list_ips(vm).
* Fixes in arguments parser.

Version 0.6:
* Script now can use all available VMs ('--vms all') and snapshots ('--snapshots all'). Use with caution and make sure you have only testing VMs.
* New function to get list of snapshots for specific VM - vm_functions.list_snapshots().
* Script updated to handle keyboard interrupts (Ctrl+C).
* Option to set random screen resolution from most common ones ('--resolution random').
* Some tweaks and fixes.

Version 0.5:
* All functions moved to external files: vm_functions (all VM control functions) and support_functions (randomize filename, calculate a hash and show links to Google/VT).
* The main routine moved to file demo.py.
* New function to get list of VMs - vm_functions.list_vms().
* Fixed logic when the file was removed before execution - script will move on to the next task without waiting for a timeout.

Version 0.4:
* Updated to use 'argparse' to parse command-line arguments.
* All of the settings now can be configured via the command line (see '--help' or examples above).

Version 0.3.1-0.3.2:
* Optionally calculate sha256 hash of a file and show links to VirusTotal and Google searches.
* Small fixes.

Version 0.3:
* Added option to enable/disable network for guest OS.

tests/test.py
import support_functions
import vm_functions
import unittest

version_good = '6.1.18r142142'
vm_good = 'ws2019'
vm_bad = 'bad'
snapshot_good = 'live'
snapshot_bad = 'bad'
file_good = '../putty.exe'
file_bad = '../bad.exe'
file_dst = 'C:\\windows\\temp\\file.exe'
user_good = 'Administrator'
pass_good = '12345678'
user_bad = 'bad'
pass_bad = 'bad'
ips_good = ['10.0.2.15', '192.168.56.113']


class TestStringMethods(unittest.TestCase):
    # vm_functions options
    vm_functions.logging.disable()
    vm_functions.vboxmanage_path = 'vboxmanage'
    vm_functions.timeout = 60

    def test01_file_info(self):
        result = support_functions.file_info(file_good)
        self.assertEqual(result[0], 0)
        self.assertEqual(result[1], 'f2d2638afb528c7476c9ee8e83ddb20e686b0b05f53f2f966fd9eb962427f8aa')
        self.assertEqual(result[2], '374fb48a959a96ce92ae0e4346763293')
        self.assertEqual(result[3], 1070)

    def test02_file_info_nonexisted(self):
        result = support_functions.file_info(file_bad)
        self.assertEqual(result, 1)

    def test03_virtualbox_version(self):
        result = vm_functions.virtualbox_version()
        self.assertEqual(result[0], 0)
        self.assertEqual(result[1], version_good)
        self.assertEqual(result[2], '')

    def test04_vm_start(self):
        result = vm_functions.vm_start(vm_good)
        self.assertEqual(result[0], 0)
        self.assertRegex(result[1], f'VM "{vm_good}" has been successfully started.')
        self.assertEqual(result[2], '')

    def test05_vm_start_running(self):
        # Note: this test expects previous test was executed too
        result = vm_functions.vm_start(vm_good)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'is already locked by a session')

    def test06_vm_start_nonexisting(self):
        result = vm_functions.vm_start(vm_bad)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'Could not find a registered machine')

    def test07_vm_upload(self):
        result = vm_functions.vm_upload(vm_good, user_good, pass_good, file_good, file_dst)
        self.assertEqual(result[0], 0)
        self.assertEqual(result[1], '')
        self.assertEqual(result[2], '')

    def test08_vm_upload_nonexisting_file(self):
        result = vm_functions.vm_upload(vm_good, user_good, pass_good, file_bad, file_dst)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'VERR_FILE_NOT_FOUND')

    def test09_vm_upload_incorrect_credentials(self):
        result = vm_functions.vm_upload(vm_good, user_bad, pass_bad, file_good, file_dst)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'The specified user was not able to logon on guest')

    def test10_vm_download_incorrect_credentials(self):
        result = vm_functions.vm_download(vm_good, user_good, pass_bad, file_bad, file_dst)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'The specified user was not able to logon on guest')

    def test11_vm_download_nonexisting_file(self):
        result = vm_functions.vm_download(vm_good, user_good, pass_good, file_dst, file_bad)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'Querying guest file information failed')

    def test12_vm_stop(self):
        result = vm_functions.vm_stop(vm_good)
        self.assertEqual(result[0], 0)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], '100%')

    def test13_vm_stop_stopped(self):
        result = vm_functions.vm_stop(vm_good)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'VBOX_E_INVALID_VM_STATE')

    def test14_vm_snapshot_restore_good(self):
        result = vm_functions.vm_snapshot_restore(vm_good, snapshot_good)
        self.assertEqual(result[0], 0)
        self.assertRegex(result[1], 'Restoring snapshot')
        self.assertRegex(result[2], '100%')

    def test15_vm_snapshot_restore_nonexisting_a(self):
        result = vm_functions.vm_snapshot_restore(vm_good, snapshot_bad)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'Could not find a snapshot')

    def test16_vm_snapshot_restore_nonexisting_b(self):
        result = vm_functions.vm_snapshot_restore(vm_bad, snapshot_bad)
        self.assertEqual(result[0], 1)
        self.assertEqual(result[1], '')
        self.assertRegex(result[2], 'Could not find a registered machine')

    def test17_list_ips(self):
        result = vm_functions.list_ips(vm_good)
        self.assertEqual(result[0], 0)
        self.assertEqual(result[1], ips_good)
        self.assertEqual(result[2], '')


if __name__ == '__main__':
    unittest.main()


demo_cli.py
import argparse
import logging
import os
import threading
import time
import http.client

script_version = '0.11'

try:
    import support_functions
    import vm_functions
except ModuleNotFoundError:
    print('Unable to import support_functions and/or vm_functions. Exiting.')
    exit(1)

# Parse command line arguments
parser = argparse.ArgumentParser(prog='vm-automation', description=f'''VirtualBox VM automation {script_version};
                                                                   https://github.com/Pernat1y/vm-automation''')

required_options = parser.add_argument_group('Required options')
required_options.add_argument('file', type=str, nargs='+', help='Path to file')
required_options.add_argument('--vms', '-v', type=str, nargs='*', required=True,
                              help='Space-separated list of VMs to use')
required_options.add_argument('--snapshots', '-s', type=str, nargs='*', required=True,
                              help='Space-separated list of snapshots to use')

main_options = parser.add_argument_group('Main options')
main_options.add_argument('--vboxmanage', default='vboxmanage', type=str, nargs='?',
                          help='Path to vboxmanage binary (default: %(default)s)')
main_options.add_argument('--check_version', action='store_true',
                          help='Check for latest VirtualBox version online (default: %(default)s)')
main_options.add_argument('--timeout', default=60, type=int, nargs='?',
                          help='Timeout in seconds for both commands and VM (default: %(default)s)')
main_options.add_argument('--delay', default=7, type=int, nargs='?',
                          help='Delay in seconds before/after starting VMs (default: %(default)s)')
main_options.add_argument('--threads', default=2, choices=range(9), type=int, nargs='?',
                          help='Number of concurrent threads to run (0=number of VMs, default: %(default)s)')
main_options.add_argument('--verbosity', default='info', choices=['debug', 'info', 'error', 'off'], type=str, nargs='?',
                          help='Log verbosity level (default: %(default)s)')
main_options.add_argument('--debug', action='store_true',
                          help='Print all messages. Alias for "--verbosity debug" (default: %(default)s)')
main_options.add_argument('--log', default=None, type=str, nargs='?',
                          help='Path to log file (default: %(default)s) (console)')
main_options.add_argument('--report', action='store_true',
                          help='Generate html report (default: %(default)s)')
main_options.add_argument('--record', action='store_true',
                          help='Record video of guest\' screen (default: %(default)s)')
main_options.add_argument('--pcap', action='store_true',
                          help='Enable recording of VM\'s traffic (default: %(default)s)')
main_options.add_argument('--memdump', action='store_true', help='Dump memory VM (default: %(default)s)')
main_options.add_argument('--no_time_sync', action='store_true',
                          help='Disable host-guest time sync for VM (default: %(default)s)')

guests_options = parser.add_argument_group('VM options')
guests_options.add_argument('--ui', default='gui', choices=['1', '0', 'gui', 'headless'], nargs='?',
                            help='Start VMs in GUI or headless mode (default: %(default)s)')
guests_options.add_argument('--login', '--user', default='user', type=str, nargs='?',
                            help='Login for guest OS (default: %(default)s)')
guests_options.add_argument('--password', default='12345678', type=str, nargs='?',
                            help='Password for guest OS (default: %(default)s)')
guests_options.add_argument('--remote_folder', default='desktop', choices=['desktop', 'downloads', 'documents', 'temp'],
                            type=str, nargs='?',
                            help='Destination folder in guest OS to place file. (default: %(default)s)')
guests_options.add_argument('--open_with', default='%windir%\\explorer.exe', type=str,
                            nargs='?', help='Absolute path to app, which will open main file (default: %(default)s)')
guests_options.add_argument('--file_args', default=None, type=str, nargs='?',
                            help='Argument to pass to the main file/executable (default: %(default)s)')
guests_options.add_argument('--network', default=None, choices=['on', 'off'], nargs='?',
                            help='State of network adapter of guest OS (default: %(default)s)')
guests_options.add_argument('--resolution', default=None, type=str, nargs='?',
                            help='Screen resolution for guest OS. Can be set to "random" (default: %(default)s)')
guests_options.add_argument('--mac', default=None, type=str, nargs='?',
                            help='Set MAC address for guest OS. Can be set to "random" (default: %(default)s)')
guests_options.add_argument('--get_file', default=None, type=str, nargs='?',
                            help='Get specific file from guest OS before stopping VM (default: %(default)s)')
guests_options.add_argument('--pre', default=None, type=str, nargs='?',
                            help='Script to run before main file (default: %(default)s)')
guests_options.add_argument('--post', default=None, type=str, nargs='?',
                            help='Script to run after main file (default: %(default)s)')

args = parser.parse_args()

# Main options
filename = args.file[0]
vms_list = args.vms
snapshots_list = args.snapshots
threads = args.threads
timeout = args.timeout
delay = args.delay
verbosity = args.verbosity
debug = args.debug
log = args.log
report = args.report
record = args.record
pcap = args.pcap
memdump = args.memdump
no_time_sync = args.no_time_sync

# vm_functions options
vm_functions.vboxmanage_path = args.vboxmanage
check_version = args.check_version
ui = args.ui
vm_functions.timeout = timeout

# VM options
vm_pre_exec = args.pre
vm_post_exec = args.post
vm_login = args.login
vm_password = args.password
remote_folder = args.remote_folder
open_with = args.open_with
file_args = args.file_args
vm_network_state = args.network
vm_resolution = args.resolution
vm_mac = args.mac
vm_get_file = args.get_file

# Some VirtualBox commands require full path to file
cwd = os.getcwd()

# Logging options
if debug:
    verbosity = 'debug'
if log == 'off':
    logging.disable()
elif verbosity in ['error', 'info', 'debug']:
    log_levels = {'error': logging.ERROR,
                  'info': logging.INFO,
                  'debug': logging.DEBUG}
    if log == 'console':
        vm_functions.logging.basicConfig(format='%(asctime)s [%(levelname)s] %(message)s', level=log_levels[verbosity])
    else:
        vm_functions.logging.basicConfig(format='%(asctime)s [%(levelname)s] %(message)s', level=log_levels[verbosity],
                                         filename=log, filemode='a')
    vm_functions.logger = logging.getLogger('vm-automation')


# Show general info
def show_info():
    logging.info(f'Script version: {script_version}')

    vbox_version = vm_functions.virtualbox_version(strip_newline=1, strip_build=1)[1]
    logging.info(f'VirtualBox version: {vbox_version}\n')

    # Check for VirtualBox version
    if check_version:
        print(f'Script version: {script_version}')
        conn = http.client.HTTPSConnection("download.virtualbox.org")
        conn.request("GET", "/virtualbox/LATEST-STABLE.TXT")
        r1 = conn.getresponse()
        if r1.status == 200:
            latest_version = r1.read().rstrip().decode("utf-8")
            if latest_version == vbox_version:
                logging.debug('Using latest VirtualBox version.')
            else:
                logging.info(f'New version of VirtualBox available: {latest_version}.')
        else:
            logging.warning('Unable to check for VirtualBox version.')

    logging.info(f'VMs: {vms_list}')
    logging.info(f'Snapshots: {snapshots_list}\n')
    result = support_functions.file_info(filename)
    if result[0] != 0:
        logging.error('Error while processing file. Exiting.')
        exit(1)
    return result[1], result[2], result[3]


# Function to take screenshot on guest OS
def take_screenshot(vm, task_name):
    screenshot_index = 1
    while screenshot_index < 10000:
        screenshot_index_zeros = str(screenshot_index).zfill(4)
        if report:
            screenshot_name_num = f'reports/{sha256}/{task_name}_{screenshot_index_zeros}.png'
        else:
            screenshot_name_num = f'{task_name}_{screenshot_index_zeros}.png'
        if os.path.isfile(screenshot_name_num):
            screenshot_index += 1
        else:
            vm_functions.vm_screenshot(vm, screenshot_name_num)
            break


# Main routines
def main_routine(vm, snapshots_list):
    for snapshot in snapshots_list:
        task_name = f'{vm}_{snapshot}'
        logging.info(f'{task_name}: Task started')

        # Create directory for report
        if report:
            os.makedirs(f'reports/{sha256}', mode=0o444, exist_ok=True)

        # Stop VM, restore snapshot
        vm_functions.vm_stop(vm, ignore_status_error=1)
        time.sleep(delay / 2)
        result = vm_functions.vm_snapshot_restore(vm, snapshot, ignore_status_error=1)
        if result[0] != 0:
            # If we were unable to restore snapshot - continue to the next snapshot/VM
            logging.error(f'Unable to restore VM "{vm}" to snapshot "{snapshot}". Skipping.')
            vm_functions.vm_stop(vm, ignore_status_error=1)
            continue
        # Change MAC address
        if vm_mac:
            vm_functions.vm_set_mac(vm, vm_mac)
        # Disable time sync
        if no_time_sync:
            vm_functions.vm_disable_time_sync(vm)
        # Dump traffic
        if pcap:
            if vm_network_state == 'off':
                logging.warning('Traffic dump enabled, but network state is set to \'off\'.')
            if report:
                pcap_file = f'{cwd}/reports/{sha256}/{vm}_{snapshot}.pcap'
            else:
                pcap_file = f'{cwd}/{vm}_{snapshot}.pcap'
            vm_functions.vm_pcap(vm, pcap_file)

        # Start VM
        time.sleep(delay / 2)
        result = vm_functions.vm_start(vm, ui)
        if result[0] != 0:
            # If we were unable to start VM - continue to the next one
            logging.error(f'Unable to start VM "{vm}". Skipping.')
            vm_functions.vm_stop(vm, ignore_status_error=1)
            continue

        # Wait for VM
        time.sleep(delay)

        # Set guest network state
        result = vm_functions.vm_network(vm, vm_network_state)
        if result[0] != 0:
            vm_functions.vm_stop(vm)
            continue

        # Set guest resolution
        vm_functions.vm_set_resolution(vm, vm_resolution)

        # Start screen recording
        if record:
            if report:
                recording_name = f'{cwd}/reports/{sha256}/{vm}_{snapshot}.webm'
            else:
                recording_name = f'{cwd}/{vm}_{snapshot}.webm'
            recording_name = support_functions.normalize_path(recording_name)
            vm_functions.vm_record(vm, recording_name)

        # Run pre exec script
        if vm_pre_exec:
            vm_functions.vm_exec(vm, vm_login, vm_password, vm_pre_exec, open_with=open_with, file_args=file_args)
            take_screenshot(vm, task_name)
        else:
            logging.debug('Pre exec is not set.')

        # Set path to file on guest OS
        remote_file_path = support_functions.randomize_filename(vm_login, filename, remote_folder)

        # Upload file to VM, check if file exist and execute
        result = vm_functions.vm_upload(vm, vm_login, vm_password, filename, remote_file_path)
        if result[0] != 0:
            take_screenshot(vm, task_name)
            vm_functions.vm_stop(vm)
            continue

        # Check if file exist on VM
        result = vm_functions.vm_file_stat(vm, vm_login, vm_password, remote_file_path)
        if result[0] != 0:
            take_screenshot(vm, task_name)
            vm_functions.vm_stop(vm)
            continue
        take_screenshot(vm, task_name)

        # Run file
        result = vm_functions.vm_exec(vm, vm_login, vm_password, remote_file_path, open_with=open_with,
                                      file_args=file_args)
        if result[0] != 0:
            take_screenshot(vm, task_name)
            vm_functions.vm_stop(vm)
            continue
        take_screenshot(vm, task_name)

        for _ in range(2):
            logging.debug(f'Waiting for {timeout / 2} seconds...')
            time.sleep(timeout / 2)
            take_screenshot(vm, task_name)

        # Check for file at the end of task
        result = vm_functions.vm_file_stat(vm, vm_login, vm_password, remote_file_path)
        if result[0] != 0:
            logging.info('Original file does not exists anymore (melted or removed by AV).')

        # Run post exec script
        if vm_post_exec:
            vm_functions.vm_exec(vm, vm_login, vm_password, vm_post_exec, open_with=open_with)
            take_screenshot(vm, task_name)
        else:
            logging.debug('Post exec is not set.')

        # Get file from guest
        if vm_get_file:
            # Normalize path and extract file name
            src_path = support_functions.normalize_path(vm_get_file)
            src_filename = os.path.basename(src_path)
            if report:
                # Place in reports directory
                dst_file = f'{cwd}/reports/{sha256}/{src_filename}'
            else:
                # Place in current dir
                dst_file = f'{cwd}/{src_filename}'
            # Download file
            vm_functions.vm_copyfrom(vm, vm_login, vm_password, src_path, dst_file)

        # Stop recording
        if record:
            vm_functions.vm_record_stop(vm)

        # Dump VM memory
        if memdump:
            if report:
                memdump_file = f'{cwd}/reports/{sha256}/{vm}_{snapshot}.dmp'
            else:
                memdump_file = f'{cwd}/{vm}_{snapshot}.dmp'
            vm_functions.vm_memdump(vm, memdump_file)

        # Stop VM
        vm_functions.vm_stop(vm)

        # Save html report as ./reports/<file_hash>/index.html
        if report:
            support_functions.html_report(vm, snapshot, filename, file_args, file_size, sha256, md5, timeout,
                                          vm_network_state)

        logging.info(f'{task_name}: Task finished')


if 'all' in vms_list:
    # If vms_list is set to 'all', obtain list of all available VMs and use them
    vms_list = vm_functions.list_vms()

# Autodetect snapshots
if 'all' in snapshots_list:
    snapshots_autodetect = True
else:
    snapshots_autodetect = False

# Number of concurrent threads
if threads == 0:
    threads = len(vms_list)
    logging.debug(f'Threads count is set to number of VMs: {threads}')
else:
    if threads > len(vms_list):
        logging.warning(f'Number of concurrent threads is larger then number of available VMs ({len(vms_list)}).')
        threads = len(vms_list)
    logging.debug(f'Threads count is set to {threads}')

# Show file information
sha256, md5, file_size = show_info()

# Start threads
for vm in vms_list:
    # Limit number of concurrent threads
    while threading.active_count() - 1 >= threads:
        time.sleep(delay / 2)

    if snapshots_autodetect:
        logging.debug('Snapshots list will be obtained from VM information.')
        snapshots_list = vm_functions.list_snapshots(vm)
        if snapshots_list[0] == 0:
            snapshots_list = snapshots_list[1]
        else:
            logging.error(f'Unable to get list of snapshots for VM "{vm}". Skipping.')
            continue

    try:
        t = threading.Thread(target=main_routine, args=(vm, snapshots_list))
        t.name = f'{vm}_{snapshots_list}'
        t.start()
        time.sleep(delay)  # Delay before starting next VM
    except (KeyboardInterrupt, SystemExit):
        raise
